File: DSA-lab/modules/01-komplexitaetsanalyse-big-o/projects/03-final/data/generate_data.py
"""Generate deterministic synthetic e-commerce events for the audit."""

from __future__ import annotations

import argparse
import csv
import random
from datetime import datetime, timedelta
from pathlib import Path

DEFAULT_SEED = 20_260_716
DEFAULT_ROWS = 8_000
DEFAULT_OUTPUT = Path(__file__).resolve().parent / "events.csv"

FIELDNAMES = (
    "event_id",
    "customer_id",
    "category",
    "amount_cents",
    "discount_cents",
    "status",
    "event_timestamp",
)

CATEGORIES = (
    "books",
    "electronics",
    "fashion",
    "garden",
    "grocery",
    "sports",
    "toys",
    "travel",
)


def generate_events(
    destination: Path,
    rows: int = DEFAULT_ROWS,
    seed: int = DEFAULT_SEED,
) -> None:
    """Write a deterministic synthetic event CSV."""
    if rows < 1:
        raise ValueError("rows must be positive")

    # A fixed local RNG makes benchmark inputs reproducible without changing
    # process-wide random state. Reproducibility is essential when comparing
    # timing changes across refactorings.
    random_generator = random.Random(seed)
    customer_count = max(25, rows // 4)
    start_time = datetime(2025, 1, 1, 0, 0)
    destination.parent.mkdir(parents=True, exist_ok=True)

    with destination.open("w", newline="", encoding="utf-8") as csv_file:
        writer = csv.DictWriter(csv_file, fieldnames=FIELDNAMES)
        writer.writeheader()

        for index in range(rows):
            status_draw = random_generator.random()
            if status_draw < 0.72:
                status = "completed"
            elif status_draw < 0.88:
                status = "pending"
            else:
                status = "cancelled"

            amount_cents = random_generator.randint(500, 75_000)
            maximum_discount = min(amount_cents // 3, 5_000)
            discount_cents = random_generator.randint(0, maximum_discount)
            timestamp = start_time + timedelta(
                minutes=random_generator.randrange(365 * 24 * 60)
            )

            writer.writerow(
                {
                    "event_id": f"EVT-{index:07d}",
                    "customer_id": (
                        f"CUST-{random_generator.randrange(customer_count):05d}"
                    ),
                    "category": random_generator.choice(CATEGORIES),
                    "amount_cents": amount_cents,
                    "discount_cents": discount_cents,
                    "status": status,
                    "event_timestamp": timestamp.isoformat(timespec="minutes"),
                }
            )


def parse_arguments() -> argparse.Namespace:
    """Parse command-line options."""
    parser = argparse.ArgumentParser(
        description="Generate deterministic synthetic e-commerce events."
    )
    parser.add_argument("--rows", type=int, default=DEFAULT_ROWS)
    parser.add_argument("--seed", type=int, default=DEFAULT_SEED)
    parser.add_argument("--output", type=Path, default=DEFAULT_OUTPUT)
    return parser.parse_args()


def main() -> None:
    """Generate a CSV from command-line settings."""
    arguments = parse_arguments()
    try:
        generate_events(arguments.output, arguments.rows, arguments.seed)
    except ValueError as error:
        raise SystemExit(f"Configuration error: {error}") from error

    print(
        f"Generated {arguments.rows} events at {arguments.output} "
        f"with seed {arguments.seed}."
    )


if __name__ == "__main__":
    main()

File: DSA-lab-java/modules/01-komplexitaetsanalyse-big-o/projects/03-final/src/main/java/lab/dsa/module01/finalproject/GenerateData.java
package lab.dsa.module01.finalproject;

import java.io.BufferedWriter;
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.List;
import java.util.Random;

/** Generates deterministic synthetic e-commerce events for the audit. */
public final class GenerateData {
    public static final long DEFAULT_SEED = 20_260_716L;
    public static final int DEFAULT_ROWS = 8_000;
    public static final Path DEFAULT_OUTPUT = Paths.get("data", "events.csv");
    private static final List<String> CATEGORIES = List.of(
            "books", "electronics", "fashion", "garden", "grocery", "sports", "toys", "travel");
    private static final DateTimeFormatter TIMESTAMP_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd'T'HH:mm");

    private GenerateData() {
    }

    public static void generateEvents(Path destination, int rows, long seed) throws IOException {
        if (rows < 1) {
            throw new IllegalArgumentException("Rows must be positive");
        }
        Random random = new Random(seed);
        int customerCount = Math.max(25, rows / 4);
        LocalDateTime startTime = LocalDateTime.of(2025, 1, 1, 0, 0);
        if (destination.getParent() != null) {
            Files.createDirectories(destination.getParent());
        }
        try (BufferedWriter writer = Files.newBufferedWriter(destination, StandardCharsets.UTF_8)) {
            writer.write("event_id,customer_id,category,amount_cents,discount_cents,status,event_timestamp");
            writer.newLine();
            for (int index = 0; index < rows; index++) {
                double statusDraw = random.nextDouble();
                String status = statusDraw < .72 ? "completed" : statusDraw < .88 ? "pending" : "cancelled";
                int amountCents = 500 + random.nextInt(74_501);
                int maximumDiscount = Math.min(amountCents / 3, 5_000);
                int discountCents = random.nextInt(maximumDiscount + 1);
                LocalDateTime timestamp = startTime.plusMinutes(random.nextInt(365 * 24 * 60));
                writer.write(String.format("EVT-%07d,CUST-%05d,%s,%d,%d,%s,%s",
                        index, random.nextInt(customerCount), CATEGORIES.get(random.nextInt(CATEGORIES.size())),
                        amountCents, discountCents, status, TIMESTAMP_FORMAT.format(timestamp)));
                writer.newLine();
            }
        }
    }

    public static void main(String[] arguments) throws IOException {
        int rows = DEFAULT_ROWS;
        long seed = DEFAULT_SEED;
        Path output = DEFAULT_OUTPUT;
        for (int index = 0; index < arguments.length; index++) {
            switch (arguments[index]) {
                case "--rows" -> rows = Integer.parseInt(arguments[++index]);
                case "--seed" -> seed = Long.parseLong(arguments[++index]);
                case "--output" -> output = Paths.get(arguments[++index]);
                default -> throw new IllegalArgumentException("Unknown option: " + arguments[index]);
            }
        }
        generateEvents(output, rows, seed);
        System.out.printf("Generated %d events at %s with seed %d.%n", rows, output, seed);
    }
}
